### ISSUE TITLE
php performance problem

### ISSUE THREAD

**Issue by cl51287** [protocolbuffers/protobuf#4227]
php version is 7.0.24, protobuf extension is v3.4.1(also in v3.5.1), 
When i test the performance use ab (nginx + php-fpm), system cpu become very high, 
```
demo.php
<?php
require 'User.php';
require 'GPBMetadata/User.php';
new User;
```
```
user.proto
syntax = "proto3";
message  User {
    repeated uint32 uids =1;
    uint32 size = 2;
}
```
ab code is 
```
ab -n 10000 -c 1000 http://127.0.0.1/demo.php
```
I found, the problem is `$pool->internalAddGeneratedFile(hex2bin("***")); in GPBMetadata/User.php`, this code take up 1ms cpu time, and every request all run this line code, so that system cpu become 30%, and if i add many `new ProtoClass`(10), the system cpu become 85%

**Comment by cl51287** [comment]
@TeBoring

**Comment by TeBoring** [comment]
Currently, the descriptor pool is built per request. For the second time you call new ProtoClass in the same request, the internalAddGeneratedPool will do nothing.

**Comment by cl51287** [comment]
@TeBoring I use release 3.5.1, but this phenomenon still appears. This phenomenon don't occur when i delete internalAddGeneratedPool this line.

**Comment by cl51287** [comment]
@TeBoring i know, the internalAddGeneratedPool  will do nothing at the second time in the same request, but next request, it will be built again, it will cause cpu become very high, when many request visit server. Can this built action cache in php-fpm process ? Because the proto does not change in the different requests.

**Comment by zvuki** [comment]
@TeBoring Can it be cached in, say, `apcu`?

**Comment by zackangelo** [comment]
@TeBoring what was the use case in mind when having the `DescriptorPool` be rebuilt every request? Is it for long-running PHP servers? 

Rebuilding the `DescriptorPool` every request is having a massively detrimental effect on performance for us because we have a ton of proto classes. Does the C extension have to do the same thing? Would you be amenable to us adding some methods to `DescriptorPool` that allow us to cache it externally?

**Comment by TeBoring** [comment]
"Because the proto does not change in the different requests" Is this guaranteed? @cl51287 
AFAIK, this is not true. Every request are free to execute any php files, which may contain conflicting protos.

**Comment by TeBoring** [comment]
@zackangelo Yes, for long-running PHP servers

**Comment by zackangelo** [comment]
@TeBoring we can guarantee in our case that the descriptors won't change. would you be opposed to us introducing a caching mechanism for the descriptor pool?

**Comment by cl51287** [comment]
@TeBoring yes, you are right, it may be change, but most of the time it does not change, as zackanglo and zvuki say , can it be cached in?

**Comment by TeBoring** [comment]
One option is add a compile flag to tell c extension to cache it. Protobuf
user needs to specify it explicitly to make sure they know what they are
doing.

**Comment by zackangelo** [comment]
@cl51287 the option doesn't exist, Paul is suggesting that as something we might build to enable this behavior. 

@TeBoring how would you feel about a compiler mode that generated code that wrote bytes to a `CodedOutputStream` directly instead of relying the descriptor to do it dynamically?

**Comment by TeBoring** [comment]
@zackangelo I don't think it's a good idea. We have tried to minimize php level code. Although your suggestion may improve the performance of pure php implementation, for cpp implementation, it will be terrible. With your proposal, most of the code will move from cpp to php.

**Comment by TeBoring** [comment]
I would suggest add a global method to tell protobuf runtime not to deactivate descriptor pool after request is finished.

**Comment by cl51287** [comment]
I found protobuf compile the descriptor pool to memory for cpp implementation, is there a method can cache this object use pure php implementation?

**Comment by TeBoring** [comment]
Not yet.

**Comment by cl51287** [comment]
Does this feature consider adding to protobuf? This compilation will take up many cpu, and there is no method to cache the result of compilation for php implementation.

**Comment by TeBoring** [comment]
Not in this release. But will add it later.

**Comment by cl51287** [comment]
thank you very much

**Comment by TeBoring** [comment]
Sorry, I haven't added it.

**Comment by zackangelo** [comment]
@TeBoring wondering if you've given this any more thought? this is still a problem that heavily affects us in production.

**Comment by TeBoring** [comment]
Sorry for late reply. I'll add it soon.

**Comment by cl51287** [comment]
> Sorry for late reply. I'll add it soon.

Will the next version add this feature?

**Comment by zackangelo** [comment]
@TeBoring just wanted to ping and see if you've had a chance to look at this :)

**Comment by TeBoring** [comment]
Fixed in https://github.com/protocolbuffers/protobuf/pull/6899
Anyone wants to try that?

**Comment by TeBoring** [comment]
To use it,
1) Use the protoc in the change to regenerate code
2) Add ```protobuf.keep_descriptor_pool_after_request=1``` into php.ini

**Comment by cl51287** [comment]
@TeBoring Thank you very much, we are waiting for a long time, try it first.

**Comment by TeBoring** [comment]
Should have been fixed by #6899
